pkg/networkservice/chains/chainstest/builder.go

```diff
@@ -0,0 +1,264 @@
+	adapter_registry "github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/registry/memory"
+	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
+)
+
+// DomainBuilder implements builder pattern for building Domain and Supplier
+type DomainBuilder struct {
+	require           *require.Assertions
+	resources         []context.CancelFunc
+	nodesCount        int
+	supplyForwarder   SupplyForwarderFunc
+	supplyNSMgr       SupplyNSMgrFunc
+	supplyRegistry    SupplyRegistryFunc
+	supplyNSE         SupplyEndpointFunc
+	supplyNSC         SupplyClientFunc
+	generateTokenFunc token.GeneratorFunc
+	ctx               context.Context
+}
+
+// NewDomainBuilder creates new DomainBuilder
+func NewDomainBuilder(t *testing.T) *DomainBuilder {
+	return &DomainBuilder{
+		nodesCount:        1,
+		require:           require.New(t),
+		supplyNSMgr:       nsmgr.NewServer,
+		supplyForwarder:   supplyDummyForwarder,
+		supplyRegistry:    supplyMemoryRegistry,
+		supplyNSE:         supplyDummyNSE,
+		supplyNSC:         client.NewClient,
+		generateTokenFunc: tokenGenerator,
+	}
+}
+
+// Build builds Domain and Supplier
+func (b *DomainBuilder) Build() (*Domain, *Supplier) {
+	ctx := b.ctx
+	if ctx == nil {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second)
+		b.resources = append(b.resources, cancel)
+	}
+	domain := &Domain{}
+	reg, u := b.newRegistry(ctx, b.supplyRegistry)
+	domain.Registry = &RegistryEntry{
+		Registry: reg,
+		URL:      u,
+	}
+	for i := 0; i < b.nodesCount; i++ {
+		var node = new(Node)
+		mgr, u := b.newNSMgr(ctx, domain.Registry.URL)
+		node.NSMgr = &NSMgrEntry{
+			Nsmgr: mgr,
+			URL:   u,
+		}
+		forwarderName := "cross-nse" + uuid.New().String()
+		forwarder, u := b.newCrossConnectNSE(ctx, forwarderName, node.NSMgr.URL)
+		node.Forwarder = &EndpointEntry{
+			Endpoint: forwarder,
+			URL:      u,
+		}
+		forwarderRegistrationClient := chain.NewNetworkServiceEndpointRegistryClient(
+			interpose_reg.NewNetworkServiceEndpointRegistryClient(),
+			adapter_registry.NetworkServiceEndpointServerToClient(node.NSMgr.NetworkServiceEndpointRegistryServer()),
+		)
+		_, err := forwarderRegistrationClient.Register(context.Background(), &registryapi.NetworkServiceEndpoint{
+			Url:  node.Forwarder.URL.String(),
+			Name: forwarderName,
+		})
+		b.require.NoError(err)
+		domain.Nodes = append(domain.Nodes, node)
+	}
+	supplier := &Supplier{
+		resources:    b.resources,
+		supplyNSC:    b.supplyNSC,
+		supplyNSE:    b.supplyNSE,
+		tokenGenFunc: b.generateTokenFunc,
+		require:      b.require,
+		dialContext:  b.dialContext,
+		ctx:          ctx,
+	}
+	b.resources = nil
```

Comment:
Why?

Reply:
This is needed if the builder will be instantly used for creating the second NSM domain.